lib/rust/profiling/src/frame_stats.rs

```diff
@@ -0,0 +1,260 @@
+//! This module defines a structure gathering statistics of the running engine. The statistics are
+//! an amazing tool for debugging what is really happening under the hood and understanding the
+//! performance characteristics.
+
+pub mod intervals;
+
+use enso_prelude::*;
+
+use js_sys::ArrayBuffer;
+use js_sys::WebAssembly::Memory;
+use serde::Deserialize;
+use serde::Serialize;
+use wasm_bindgen::JsCast;
+
+
+
+// =============
+// === Stats ===
+// =============
+
+/// Structure containing all the gathered stats.
+#[derive(Debug, Clone, CloneRef)]
+pub struct Stats {
+    rc: Rc<RefCell<StatsData>>,
+}
+
+impl Default for Stats {
+    fn default() -> Self {
+        let rc = Rc::new(RefCell::new(default()));
+        Self { rc }
+    }
+}
+
+impl Stats {
+    /// Starts tracking data for a new animation frame.
+    pub fn begin_frame(&self, time: f64) {
+        self.rc.borrow_mut().begin_frame(time);
+    }
+
+    /// Ends tracking data for the current animation frame.
+    pub fn end_frame(&self, time: f64) {
+        self.rc.borrow_mut().end_frame(time);
+    }
+
+    /// Resets the per-frame statistics.
+    pub fn reset_per_frame_statistics(&self) {
+        self.rc.borrow_mut().reset_per_frame_statistics();
+    }
+
+    /// Returns a read-only reference to the underlying raw data.
+    pub fn data(&self) -> Ref<StatsData> {
+        self.rc.borrow()
+    }
```

Comment:
Thanks for the further explanation and discussion on a call on previous week!

Based on what we discussed, I understood the ideal option here would be to `impl Deref` to provide access `StatsData`; however, IIUC that won't work, because implementing the `Deref` trait requires returing a reference (`fn deref(&self) -> &Self::Target`), and due to `RefCell::borrow()` semantics (returning a temporary `Ref<StatsData>` object), we can't get a `&StatsData` reference that would outlive the `deref()` call.

Due do that, I decided to instead make a copy of `StatsData` here. This copy will be done only when profiling is enabled, and only once per frame. The "once per frame" is not necessarily trivial here given the amount of stats in the struct, but `StatsData` is `Copy` so it should still be relatively cheap to do that, so I would assume this should hopefully not be a problem. However, if it is, please see below.

An alternative I considered was to rename this fn from `data` to `borrow` to make it more explicit, and to add a comment explaining that no other methods on `Stats` may be used while the borrow (i.e. the returned `Ref` object) is alive. This should be cheaper than the `snapshot` described above (not needing the `StatsData` copy per frame); however, it would incur some extra complexity of having to manually track lifetimes to make sure that we don't accidentally trigger a panic by calling some other method of `Stats` (namely, any one of them that does a `borrow_mut()` internally).

If I should go with the presumably cheaper `borrow` alternative described above, or if you have some better suggestion as to how I could resolve it, that I missed, I'd be grateful if you could let me know!